src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -257,91 +317,103 @@ public boolean isFingerprintArtifacts() {
-
-            if (Hudson.getInstance().getPlugin("maven-plugin") != null && (src instanceof MavenModuleSetBuild) ) { 
-            // use classes in the "maven-plugin" plugin as might not be installed
-                // Copy artifacts from the build (ArchiveArtifacts build step)
-                boolean ok = perform(src, build, expandedFilter, expandedExcludes, targetDir, baseTargetDir, copier, console);
-                // Copy artifacts from all modules of this Maven build (automatic archiving)
-                for (Iterator<MavenBuild> it = ((MavenModuleSetBuild)src).getModuleLastBuilds().values().iterator(); it.hasNext(); ) {
-                    // for(Run r: ....values()) causes upcasting and loading MavenBuild compiled with jdk 1.6.
-                    // SEE https://wiki.jenkins-ci.org/display/JENKINS/Tips+for+optional+dependencies for details.
-                    Run<?,?> r = it.next();
-                    ok |= perform(r, build, expandedFilter, expandedExcludes, targetDir, baseTargetDir, copier, console);
-                }
-                return ok;
-            } else if (src instanceof MatrixBuild) {
-                boolean ok = false;
-                // Copy artifacts from all configurations of this matrix build
-                // Use MatrixBuild.getExactRuns if available
-                for (Run r : ((MatrixBuild) src).getExactRuns())
-                    // Use subdir of targetDir with configuration name (like "jdk=java6u20")
-                    ok |= perform(r, build, expandedFilter, expandedExcludes, targetDir.child(r.getParent().getName()),
-                                  baseTargetDir, copier, console);
-                return ok;
-            } else {
-                return perform(src, build, expandedFilter, expandedExcludes, targetDir, baseTargetDir, copier, console);
+        } else {
+            if (!perform(src, build, expandedFilter, expandedExcludes, targetDir, baseTargetDir, copier, console)) {
+                throw new AbortException(Messages.CopyArtifact_FailedToCopy(expandedProject, expandedFilter));
             }
         }
-        catch (IOException ex) {
-            Util.displayIOException(ex, listener);
-            ex.printStackTrace(listener.error(
-                    Messages.CopyArtifact_FailedToCopy(expandedProject, expandedFilter)));
-            return false;
-        }
     }
 
-    private boolean canReadFrom(Job<?, ?> job, AbstractBuild<?, ?> build) {
-        if ((job instanceof AbstractProject) && CopyArtifactPermissionProperty.canCopyArtifact(
-                build.getProject().getRootProject(),
-                ((AbstractProject<?,?>)job).getRootProject()
-        )) {
+    private boolean canReadFrom(Job<?, ?> job, Run<?, ?> build) {
+        Job<?, ?> fromJob = job;
+        Job<?, ?> toJob = build.getParent();
+
+        if ((job instanceof AbstractProject) && (build instanceof AbstractBuild) &&
```

Comment:
Right, probably not needed. Have removed and am running tests.

While `canCopyArtifact` can take `Job` for both of its args, I had in my head (wrongly I'm sure) that the original implementor only wanted to call that method for `AbstractProject` types (for some reason) and otherwise move on to the code after. In hindsight that doesn't seem to make sense.